### PR TITLE
 Add new loss layer for binary loss per pixel

### PR DESCRIPTION
**Pull request by reunanen**
- Add new loss layer for binary loss per pixel (say, for completeness)
- In the instance-segmentation example, resolve possibly overlapping instances using the per-pixel output intensity (and not result rect size as previously); while at it, also make the instance output look "soft":
image

### CODE REVIEW

File: dlib/cuda/cuda_dlib.cu [hunk 1]
```diff
@@ -1624,6 +1624,48 @@ namespace dlib
             }
         }
 
+    // ----------------------------------------------------------------------------------------
+
+        __device__ float cuda_log1pexp(float x)
```
Comment (reunanen) [inline dlib/cuda/cuda_dlib.cu]: This implementation was copied from [utilities.h](https://github.com/davisking/dlib/blob/d766f5e82e650fb2b24bbbb8c5d9322a9fcef36c/dlib/dnn/utilities.h#L16-L29), but I'm not sure if the magic numbers should be different for `float` input?

File: examples/dnn_instance_segmentation_ex.cpp [hunk 2]
```diff
@@ -16,7 +16,7 @@
        ./dnn_instance_segmentation_ex /path/to/VOC2012-or-other-images
 
     An alternative to steps 2-4 above is to download a pre-trained network
-    from here: http://dlib.net/files/instance_segmentation_voc2012net.dnn
+    from here: http://dlib.net/files/instance_segmentation_voc2012net_v2.dnn
```
Comment (reunanen) [inline examples/dnn_instance_segmentation_ex.cpp]: For the time being at least, this file is available here:
https://www.dropbox.com/s/t7gi4lfw6mrx10v/instance_segmentation_voc2012net_v2.dnn

File: dlib/test/dnn.cpp [hunk 3]
```diff
@@ -3396,6 +3592,9 @@ namespace
             test_multioutput_linear_regression();
             test_simple_autoencoder();
             test_loss_mean_squared_per_channel_and_pixel();
+            test_loss_binary_log_per_pixel_learned_params_on_trivial_two_pixel_task();
+            test_loss_binary_log_per_pixel_outputs_on_trivial_task();
+            test_loss_binary_log_per_pixel_with_noise_and_pixels_to_ignore();
```
Comment (reunanen) [inline dlib/test/dnn.cpp]: Not sure if all these tests are really useful...

File: examples/dnn_instance_segmentation_ex.cpp [hunk 4]
```diff
@@ -71,26 +71,21 @@ int main(int argc, char** argv) try
     {
         // Load the input image.
         load_image(input_image, file.full_name());
-
-        // Draw largest objects last
```
Comment (reunanen) [inline examples/dnn_instance_segmentation_ex.cpp]: No need to sort the detections anymore.

File: dlib/cuda/cuda_dlib.cu (hunk 1 above)
Comment (reunanen) [reply dlib/cuda/cuda_dlib.cu]: Looking at this, I guess the answer is **yes**: https://github.com/JuliaStats/StatsFuns.jl/blob/0beb42294785787ac58fec247cd5d225a99e6d0f/src/basicfuns.jl#L67

Btw, the first Google hit was [this question](https://cs.stackexchange.com/questions/110798/numerically-stable-log1pexp-calculation) :)

Comment (reunanen) [reply dlib/cuda/cuda_dlib.cu]: Fixed.

File: dlib/cuda/cuda_dlib.cu [hunk 5]
```diff
@@ -1663,6 +1705,30 @@ namespace dlib
...
+            sigmoid(gradient, subnetwork_output);
+
+            // The loss we output is the average loss over the mini-batch, and also over each element of the matrix output.
+            const double scale = 1.0 / (subnetwork_output.num_samples() * subnetwork_output.nr() * subnetwork_output.nc());
+
+            launch_kernel(_cuda_compute_loss_binary_log_per_pixel, max_jobs(gradient.size()),
+                loss_cuda_work_buffer, gradient.device(), truth_buffer, subnetwork_output.device(), gradient.size(), scale);
+
+            float floss;
+            CHECK_CUDA(cudaMemcpy(&floss, loss_cuda_work_buffer, sizeof(float), cudaMemcpyDefault));
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.cu]: use `dlib::cuda::memcpy()` instead of `cudaMemcpy()`

File: dlib/cuda/cuda_dlib.h [hunk 6]
```diff
@@ -412,6 +412,62 @@ namespace dlib
 
     // ----------------------------------------------------------------------------------------
 
+        class compute_loss_binary_log_per_pixel
+        {
+            /*!
+                The point of this class is to compute the loss computed by
+                loss_binary_log_per_pixel, but to do so with CUDA.
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.h]: Should reference `loss_binary_log_per_pixel_` not `loss_binary_log_per_pixel`

File: dlib/cuda/cuda_dlib.h [hunk 7]
```diff
@@ -412,6 +412,62 @@ namespace dlib
...
+                double& loss
+            ) const
+            {
+                const size_t bytes_per_plane = subnetwork_output.nr()*subnetwork_output.nc()*sizeof(float);
+                // Allocate a cuda buffer to store all the truth images and also one float
+                // for the scalar loss output.
+                buf = device_global_buffer(subnetwork_output.num_samples()*bytes_per_plane + sizeof(float));
+
+                cuda_data_void_ptr loss_buf = buf;
+                buf = buf+sizeof(float);
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.h]: I just realized there isn't any method to cast a `cuda_data_void_ptr` to a `cuda_data_ptr<float>`, which is really what should happen here :|

**Comment by davisking** [comment]
Got to take my kid to school.  I'll look more at this later.

File: dlib/cuda/cuda_dlib.cu (hunk 5 above)
Comment (reunanen) [reply dlib/cuda/cuda_dlib.cu]: Fixed – also at the source where this was basically copied from.

File: dlib/cuda/cuda_dlib.h (hunk 6 above)
Comment (reunanen) [reply dlib/cuda/cuda_dlib.h]: Fixed – also at the source where this was copied from.

File: dlib/test/dnn.cpp [hunk 8]
```diff
@@ -2555,6 +2555,202 @@ namespace
...
+        trainer.set_max_num_epochs(200);
+        trainer.train(x, y);
+
+        // The learning task is easy, so the net should have no problem
+        // getting all the outputs right.
+        const auto response = net(x);
+        for (int ii = 0; ii < num_samples; ++ii)
+            for (int jj = 0; jj < output_height; ++jj)
+                for (int kk = 0; kk < output_width; ++kk)
+                    DLIB_TEST((response[ii](jj,kk) > 0) == (y[ii](jj,kk) > 0));
```
Comment (davisking) [inline dlib/test/dnn.cpp]: I'm getting a failure here when I run this test with dlib built to use cuda.  Is it working for you?

File: dlib/cuda/cuda_dlib.h (hunk 7 above)
Comment (davisking) [reply dlib/cuda/cuda_dlib.h]: I just pushed `static_pointer_cast()`.  So you can do this now:
```
cuda_data_ptr<float> loss_buf = static_pointer_cast<float>(buf, 1);
```
and `loss_buf.get() == buf.get()` and `loss_buf.size() == 1`.  Then you can use the nicely typed loss_buf rather than void pointers.

File: dlib/test/dnn.cpp (hunk 8 above)
Comment (reunanen) [reply dlib/test/dnn.cpp]: Yes, it works on my machine (CUDA build).

Tried dozens of times now, but couldn't get a failure.

Doubled the number of epochs to train. Does that help?

File: dlib/cuda/cuda_dlib.h (hunk 7 above)
Comment (reunanen) [reply dlib/cuda/cuda_dlib.h]: This should look better now.

File: dlib/test/dnn.cpp (hunk 8 above)
Comment (arrufat) [reply dlib/test/dnn.cpp]: I am also getting an error with 400 epochs. Doubling it works for me, but also setting it 500.

Comment (reunanen) [reply dlib/test/dnn.cpp]: Thanks @arrufat. On my machine it worked with 200, and the 400 was already a doubled-up number. Now increased further to 800.

File: dlib/cuda/cuda_dlib.h [hunk 9]
```diff
@@ -424,11 +424,67 @@ namespace dlib
...
+                }
+
+                do_work(loss_buf, static_cast<float*>(buf.data()), subnetwork_output, gradient, loss);
+            }
+
+        private:
+
+            static void do_work(
+                cuda_data_ptr<float> loss_cuda_work_buffer,
+                const float* truth_buffer,
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.h]: Can `truth_buffer` be a cuda_data_ptr too?  The way the code appears suggests `truth_buffer` is a host pointer.

File: dlib/cuda/cuda_dlib.cu [hunk 10]
```diff
@@ -1720,10 +1762,34 @@ namespace dlib
...
+            float floss;
+            dlib::cuda::memcpy(&floss, loss_cuda_work_buffer);
+            loss = scale*floss;
+        }
 
         void compute_loss_multiclass_log_per_pixel::
         do_work(
-            float* loss_cuda_work_buffer,
+            cuda_data_ptr<float> loss_cuda_work_buffer,
             const uint16_t* truth_buffer,
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.cu]: Can `truth_buffer` be a cuda_data_ptr too?  The way the code appears suggests `truth_buffer` is a host pointer.

File: dlib/cuda/cuda_dlib.cu [hunk 11]
```diff
@@ -1720,10 +1762,34 @@ namespace dlib
...
+
+            float floss;
+            dlib::cuda::memcpy(&floss, loss_cuda_work_buffer);
+            loss = scale*floss;
+        }
 
         void compute_loss_multiclass_log_per_pixel::
         do_work(
-            float* loss_cuda_work_buffer,
+            cuda_data_ptr<float> loss_cuda_work_buffer,
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.cu]: Can `truth_buffer` be a cuda_data_ptr too?  The way the code appears suggests `truth_buffer` is a host pointer, since `loss_cuda_work_buffer` is so "hey I'm a cuda buffer!" but `truth_buffer` isn't.  

Also, `loss_cuda_work_buffer` should be named something like `loss_work_buffer`, the type is making it clear it's a cuda buffer.

File: dlib/cuda/cuda_dlib.h [hunk 12]
```diff
@@ -464,13 +520,13 @@ namespace dlib
...
 
-                do_work(static_cast<float*>(loss_buf.data()), static_cast<uint16_t*>(buf.data()), subnetwork_output, gradient, loss);
+                do_work(loss_buf, static_cast<uint16_t*>(buf.data()), subnetwork_output, gradient, loss);
             }
 
         private:
 
             static void do_work(
-                float* loss_cuda_work_buffer,
+                cuda_data_ptr<float> loss_cuda_work_buffer,
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.h]: Same comments here about cuda naming and `truth_buffer` (i.e. it should be a cuda_data_ptr too so the arguments are consistent)

File: dlib/test/dnn.cpp (hunk 8 above)
Comment (davisking) [reply dlib/test/dnn.cpp]: Cool, working for me now too :)

File: dlib/cuda/cuda_dlib.cu (hunk 11 above)
Comment (reunanen) [reply dlib/cuda/cuda_dlib.cu]: > Can truth_buffer be a cuda_data_ptr too?

I did think about it, but wasn't sure if I can make it `const` (meaning that the data pointed to won't be changed – obviously the pointer object argument itself can be made `const`).

Comment (reunanen) [reply dlib/cuda/cuda_dlib.cu]: > Also, `loss_cuda_work_buffer` should be named something like `loss_work_buffer`, the type is making it clear it's a cuda buffer.

Fixed.

Comment (reunanen) [reply dlib/cuda/cuda_dlib.cu]: >> Can truth_buffer be a cuda_data_ptr too?

> I did think about it, but wasn't sure if I can make it const

Looking at how [these](https://github.com/davisking/dlib/blob/6bdd289f734839200226de525b78046268905cff/dlib/cuda/cuda_data_ptr.h#L68-L71) are defined, I guess constness is preserved if I pass a `const cuda_data_ptr<T>&`. (?)

**Comment by davisking** [comment]
cuda_data_ptr<const T> should do the trick. Since then the constness will stick through any copying.

**Comment by reunanen** [comment]
> cuda_data_ptr<const T> should do the trick.

Right. Thought we can't do this, because we need to populate the truth buffer as well – but of course that need not happen through this same pointer object. :)

**Comment by davisking** [comment]
Thanks for another nice PR :)
